### PR TITLE
getTimeline: historical ranges, author filtering, and day-granular coverage

### LINKED ISSUE #41: `getTimeline`: support historical ranges, author filtering, and pagination
`getTimeline` currently works for small, recent, site-wide activity queries, but it cannot answer historical or contributor-specific questions reliably. Please add explicit date-range filtering, server-side author filtering, and pagination.

This is needed to replace clients that query the WordPress Trac timeline directly. Without it, questions such as “What did jonsurrell work on last week?” can return incomplete results, while questions such as “What did saxmatt work on in January 2005?” cannot be answered at all.

## Current behavior

The current input schema exposes:

- `days`: relative lookback from now, default 7, maximum 30.
- `limit`: maximum number of global events, default 20, maximum 100.

It does not expose:

- An arbitrary start or end date.
- An author filter.
- A page, offset, or cursor.
- Response metadata indicating whether matching activity was truncated.

Client-side author filtering is not sufficient. If more than 100 global events occurred in the requested window, an author's events may fall outside the returned set even when the client requests the maximum limit.

## Requested interface

Please extend `getTimeline` with:

- `from` and `to` ISO-8601 dates for arbitrary historical ranges.
- `author`, accepting one username or a list of usernames.
- Cursor- or page-based pagination.
- Result metadata such as `returnedEvents`, `totalEvents` or `hasMore`, and `nextCursor`/`nextPage`.

The existing `days` and `limit` arguments should remain supported for backward compatibility. If `days` cannot be combined with `from`/`to`, the schema and validation error should state that explicitly.

Example request:

```json
{
  "from": "2005-01-01",
  "to": "2005-01-31",
  "author": "saxmatt",
  "limit": 100
}
```

## Acceptance criteria

- A caller can retrieve activity for an arbitrary historical date range, including dates more than 30 days ago.
- A caller can filter by one or more Trac usernames on the server.
- A contributor-specific query remains complete when the unfiltered window contains more than 100 events.
- A caller can paginate through all matches without omissions or duplicates.
- The response makes truncation or additional pages explicit.
- Existing `{ "days": 7, "limit": 20 }` calls retain their current behavior.
- The tool description documents the date, author, pagination, and result-limit semantics.

## Why this belongs in the server

The server has access to the complete Trac result set before applying output limits. Only it can filter by author and paginate without silently losing matches. Client-side filtering of an already capped global timeline cannot provide complete or trustworthy answers.

### PR DESCRIPTION
**Pull request by sirreal**
Fixes #41.

## What this does

`getTimeline` now answers historical and contributor-specific questions:

- **`from`/`to` ISO dates** reach arbitrary history, at most 90 days per request — core.trac clamps `daysback` at Trac's default `max_daysback` of 90 and silently drops anything older, so longer periods are walked as adjacent windows.
- **`author`** takes one Trac username or up to ten. It maps to Trac's server-side `authors` filter, which applies before truncation, so a contributor's events stay complete inside a busy window.
- **Coverage is reported in whole UTC days.** Responses carry `requested`, `covered`, `complete`, and — when incomplete — `continueWith`, a ready-made arguments object for the remainder that preserves the accepted `author` and effective `limit`. `limit` is advisory and rounds down to day boundaries; the newest day always comes back whole.

## What happened to the pagination

An earlier revision of this branch exposed the timeline as `page`/`limit` with `hasMore`/`nextPage`. That was the wrong design, and review demonstrated it against the live server three ways:

- Trac cannot seek to an offset, so page N refetched pages 1 through N−1 and threw them away. Quadratic load on a shared community server is not an acceptable price for a convenience field.
- Offsets index into a live, newest-first feed. One commit landing between two requests shifts every offset — a duplicated or dropped event at each page boundary, under documentation that promised neither. A design that makes its own documentation false is broken, however clean the code.
- `nextPage` could point past the server's own `page × limit ≤ 1000` validation cap, so a client following the documented protocol dead-ended on an error we built for it.

Three structural failures from one decision means the decision goes, not the symptoms. Day-granular coverage is what the upstream actually supports: Trac's timeline seeks by calendar day and by nothing else. The server makes one bounded fetch, post-filters events into the requested window by UTC date, and drops a possibly half-fetched oldest day, so every day it reports is a day it saw whole. Continuation windows always end in the past, and the past does not change: walking them cannot gap or duplicate, by construction rather than by promise.

An opaque cursor was considered and rejected: it would have to encode intra-day position and resume by refetching the whole boundary day, rebuilding on every call what a day window states outright. [docs/timeline-pagination.md](https://github.com/sirreal/trac-mcp/blob/update/timeline-history-authors-pagination/docs/timeline-pagination.md) records the reasoning; read it before adding pagination to anything else in this server.

## Also fixed, from the same review

- The advertised JSON schema and the Zod schema are now generated from the same constants and compared structurally by a test. The advertised `days: {default: 7}` was worse than cosmetic: a client that materializes defaults would send `days: 7` on every call, and every historical query would be rejected as `days` combined with `from`/`to`.
- The clock is read once per request, so validation and window resolution cannot disagree across UTC midnight.
- A same-day window no longer leaks the previous day (Trac clamps `daysback=0` to 1; post-filtering removes what that adds).
- A future `to` is rejected with `-32602` instead of returning a silently empty success.
- Author names no longer admit trailing or doubled spaces, which Trac quotes verbatim and matches against nobody.
- Dates in years below 1000 validate correctly (`Date.UTC` maps years 0–99 to 1900–1999; the roundtrip now uses `setUTCFullYear`).

## Verification

- `pnpm check` passes: types, Biome, 50 Vitest tests, both Worker dry-run builds.
- Live against `pnpm dev` per `docs/testing.md`: the January 2005 `saxmatt` fixture walks to `complete: true` at both `limit: 20` (5 steps) and `limit: 5` (13 steps); both walks tile the month with no gap or overlap and total the same 89 events. `{"from":"2005-01-31","to":"2005-01-31"}` returns exactly the 42 events of that day — the previous-day leak is gone. All five invalid-input cases fail with `-32602` before any upstream request.
- `docs/smoke-test.md` now walks a range to termination. The old checklist promised a final page and never ran one; this one cannot pass without reaching the end.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -911,15 +1003,161 @@ async function fetchTracInfo(type: TracInfoType): Promise<string[]> {
...
+    const oldestDay = events.at(-1)?.day;
+    if (oldestDay === undefined || oldestDay === events[0]?.day) {
+      break;
+    }
+    events = events.filter((event) => event.day !== oldestDay);
+    coveredFrom = shiftTimelineDay(oldestDay, 1);
+  }
+
+  const covered: TimelineWindow = { from: coveredFrom, to: requested.to };
+  const continueWith =
```
Comment (lancewillett) [inline src/index.ts]: `continueWith` drops the effective `author` and `limit`, so it does not resume the same query.

I submitted the returned object from an `author: "saxmatt", limit: 20` request unchanged. The next call returned 90 events from seven authors.

This conflicts with the README’s instruction to send `continueWith` back to `getTimeline`. It also breaks the requirement to retrieve all filtered matches without omissions or unrelated results.

Could `continueWith` preserve the accepted `author` input and effective `limit`? Please add a test that submits the returned object unchanged.

---
<sub>Adversarial review · <code>gpt-5</code></sub>

**Review by lancewillett — CHANGES_REQUESTED** [review]
Day-granular coverage fits Trac’s upstream behavior better than offset pagination. The live filtered walk completed in five requests with 89 events when I manually preserved the original arguments.

The returned continuation currently loses those arguments and changes the query. I’m requesting that fix before merge.

---
<sub>Adversarial review · <code>gpt-5</code></sub>

**Review by sirreal — COMMENTED** [review]
Adversarial review found five actionable timeline-contract gaps:

1. **High — A 500-event day cannot be traversed, and `covered` includes a known-incomplete day.** When a day fills the fetch cap, the response reports that day inside `covered`, despite the documented invariant that `covered` contains only complete days. A single-day request gets no `continueWith`, so events after the newest 500 are permanently unreachable. Represent the incomplete day outside `covered`, and provide a resumable mechanism or explicitly scope the API as unable to enumerate every match.

2. **High — Existing `{ days, limit }` callers are not backward compatible.** Previously `limit` was sent upstream as `max`, `results.length <= limit`, and the response included `totalEvents` and `daysBack`. The new implementation always fetches 500, makes `limit` advisory, and removes both fields. Thus `{ days: 1, limit: 1 }` can return hundreds of events and legacy consumers lose fields they may read. Preserve the old behavior/fields for legacy calls or make this an explicitly versioned contract change.

3. **Medium — Malformed RSS dates are silently dropped while coverage is reported complete.** A feed containing one valid item and one `pubDate` of `not-a-date` reproduced a result containing only the valid event with `complete: true`. Since upstream responses are untrusted, an invalid dated item should make coverage incomplete or produce a tool error.

4. **Medium — The accepted date domain is broader than Trac can safely serve.** Live Trac silently substituted current activity for year `0000` and returned an overflow error for year `0001`; the current four-digit date schema accepts both. This can become an empty-but-`complete` historical response after post-filtering. Constrain dates to Trac’s supported range or detect upstream fallback/errors reliably.

5. **Low — `tools/list` advertises fractional numbers that runtime validation rejects.** Runtime requires integer `days` and `limit`, but discovery advertises `type: "number"`. Advertise `type: "integer"` and extend the schema-alignment projection so it includes Zod’s integer constraint.

The existing 50-test suite, TypeScript, and Biome checks pass. Targeted reproductions confirmed the malformed-date and compatibility findings; live Trac checks confirmed the early-year behavior. Live checks also ruled out the suspected 90-day off-by-one: Trac clamps `daysback=91` to 90 while retaining the inclusive start day.

---
<sub>Adversarial review · Codex subagent</sub>

**Comment by sirreal** [comment]
### Implementation and follow-up scope

The blocking fixes are now in `f18b1d8`:

- Calls that use only `days` and `limit` keep the original recent-timeline request and response contract.
- A malformed or missing RSS `pubDate` returns a tool error instead of a false complete result.
- Historical dates start at the verified `2005-01-01` boundary.
- The advertised `days` and `limit` schemas now match the runtime integer constraints.
- A day that reaches the 500-event fetch limit is outside `covered` and has explicit `terminalTruncation` metadata.

If we accept the bounded contract in the separate scope question, I propose this follow-up work:

- Investigate whether Trac has an intra-day cursor or another official read endpoint that can traverse a day with more than 500 events.
- Define a richer partial-result contract for malformed RSS dates. This PR returns a tool error instead of silently omitting an event.
- Expand the supported historical date range only after live Trac checks show that the earlier range works.

The intra-day traversal work is follow-up only if we revise the current “all matches without omissions” acceptance criterion. If we keep that criterion, this work is required before merge.

**Comment by sirreal** [comment]
### Scope question

Wait — I do not understand which `getTimeline` contract this PR will ship.

Issue #41 says that `getTimeline` must return all matches without omissions. The Trac timeline is fetched by whole UTC day. One day can reach the 500-event fetch limit. Trac does not provide an intra-day cursor.

This means that `continueWith` cannot traverse the missing events in that day.

Please choose one scope:

1. Revise issue #41. Define an honest bounded contract. `getTimeline` returns all matches in the `covered` window. A day that reaches the 500-event fetch limit is a terminal incomplete day. That day is not in the `covered` window. `complete` is `false`. The result reports the terminal incomplete day clearly.

2. Keep the current “all matches without omissions” acceptance criterion. In that case, this PR needs a new mechanism that can traverse all events in the terminal day before merge.

I recommend option 1. It states exactly what the server can prove from the Trac timeline response. It does not claim that the `requested` window is complete when Trac did not provide all events.

**Comment by lancewillett** [comment]
Decision: option 1, the bounded contract. Trac's timeline has no intra-day cursor, so a day that fills the 500-event fetch is reported outside `covered`, with `complete: false` and `terminalTruncation`, as f18b1d8 already does. I restated the acceptance criteria on #41 to match.

I merged trunk into this branch in 96d8262 rather than asking for a rebase. The multi-instance change from #53 now threads the connected Trac instance through `fetchTimeline`, `fetchLegacyTimeline`, `buildTimelineResult`, and `getTimelineForChatGPT`, replacing the literal Core timeline URLs. f48abf0 keeps the ChatGPT compatibility search on the 20-event recent fetch. dd31177 adds an HTML-instead-of-RSS guard, pins the connected instance on the days/limit and ChatGPT paths, and replaces the schema-alignment test that walked Zod's `_def` tree with a literal comparison, so it survives the zod 4 bump in #69. `pnpm check` passes with 114 tests under zod 3 and zod 4.

**Review by lancewillett — APPROVED** [review]
Re-reviewed after f18b1d8 and the trunk merge.

- The August 10 blocker is closed: `continueWith` carries `author` and the effective `limit`, and the round-trip test replays it against a second mocked fetch.
- The five adversarial findings from August 13 are each addressed in f18b1d8. Legacy `{days, limit}` calls keep `totalEvents` and `daysBack` with `max` sent upstream. A malformed `pubDate` is a tool error. Dates start at 2005-01-01. `days` and `limit` advertise `integer`. A capped day sits outside `covered` with `terminalTruncation`.
- Three maintainer commits on the branch: 96d8262 merges trunk and threads the instance through the timeline code, with a test that `/mcp/meta` stays on Meta. f48abf0 keeps the ChatGPT compatibility search on the 20-event recent fetch instead of the 500-event coverage fetch. dd31177 makes an HTML body where RSS was expected an `upstream_error` instead of an empty `complete: true`, and pins the connected instance on the days/limit and ChatGPT paths. It also replaces the schema-alignment test that walked Zod's `_def` tree, which fails on zod 4 (#69), with a literal comparison plus runtime boundary rows. Verified green under zod 3 and zod 4.
- `pnpm check` passes with 114 tests. CI is green.

Approving. Note for merge order: #62 changes the `parseRssItems` signature, so whichever of #42 and #62 lands second needs one more trunk merge.

---
<sub>Adversarial review · <code>claude-fable-5-1</code></sub>
